### PR TITLE
Add constant folding optimization

### PR DESCRIPTION
**Pull request by larsk21**
Add optimization that finds constant values for value nodes where possible and replaces these node with constant nodes in the Firm graph.

### CODE REVIEW

**Comment by larsk21** [comment]
To do:
- [ ] add tests (requires working graph construction, especially for expressions)
- [ ] implement remaining operations on constant values
- [ ] figure out how to handle control flow (if we can do that with constant folding)

File: src/main/java/edu/kit/compiler/optimizations/ConstantOptimization.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * Transform the given node with the given associated lattice element if
+     * that element is constant.
+     */
+    private void transform(Node node, TargetValueLatticeElement value) {
+        if (value.isConstant()) {
+            Node constantNode = graph.newConst(value.getValue());
+            Graph.exchange(node, constantNode);
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/ConstantOptimization.java:40]: I think your may run into issues here. For example _Div_ nodes are of mode `T` and can't be replaced directly with constants (_Mod_ is the same). I don't know if there is a built-in way of doing this, but you could use back edges to replace the memory and result projections.

Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/ConstantOptimization.java:40]: Yeah, known issue, will fix it.

File: src/main/java/edu/kit/compiler/JavaEasyCompiler.java [hunk 2]
```diff
@@ -184,6 +190,12 @@ private static Result compileFirm(String filePath, Logger logger) {
             ast.accept(irv);
             Lower.lower(irv.getTypeMapper());
 
+            for (Optimization optimization : optimizations) {
+                for (Graph graph : Program.getGraphs()) {
+                    optimization.optimize(graph);
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/JavaEasyCompiler.java:195]: If we swap the order of the two loops and additionally let an optimization return whether anything changed or not, we could do a generic fixpoint iteration that includes all optimizations

**Review by N-Maas — COMMENTED** [review]
The functionality looks good to me and all tests from the central mjtest repo succeed :+1:

File: src/main/java/edu/kit/compiler/io/StackWorklist.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * Create a new StackWorklist.
+     */
+    public StackWorklist() {
+        stack = new Stack<>();
+        uniqueElements = true;
+    }
+
+    private Stack<T> stack;
```
[the remark this answers is not in the document]
Comment (larsk21) [reply src/main/java/edu/kit/compiler/io/StackWorklist.java:21]: Interesting, I didn't know that!

**Review by N-Maas — APPROVED** [review]

**Pull request merged**